### PR TITLE
Initial ObjectFormat implementation

### PR DESCRIPTION
**Pull request by 0xdaryl**
* Initial ObjectFormat and FunctionCallData definitions
* Create X86 specialization of FunctionCallData
* Initial AMD64 implementation of JitCodeRWXObjectFormat
* Initial AMD64 implementation of JitCodeRXObjectFormat

Issue: #5569

### CODE REVIEW

File: compiler/x/CMakeLists.txt
Comment (DanHeidinga) [inline compiler/x/CMakeLists.txt:72]: nitpick - tabs vs spaces

Comment (0xdaryl) [reply compiler/x/CMakeLists.txt]: Fixed.  Thanks.

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Comment by 0xdaryl** [comment]
@r30shah @fjeremic : I would appreciate a review from you please.

This PR simply introduces the API and an implementation on AMD64 for the RWX and RX object formats.  It does not provide a 32-bit object format (that will be coming later).  I implemented it slightly differently from the design in #5569 in that it only provides an API for a generic "function call" rather than separate APIs for "global functions" and "code cache functions". I found there was too much ambiguity in deciding when to choose one over the other that I went with a single API instead.

This work needs to be introduced in stages to prevent downstream build breakages.  The next PR after this one in OMR will use the API on AMD64 defined in this PR.

File: compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp
Comment (0xdaryl) [inline compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:96]: **Note to reviewers:**  this implementation uses `allocateCodeMemory`.   This is only a temporary solution until #5634 is merged, in which case it will switch to allocating such memory from CCData.

**Comment by 0xdaryl** [comment]
Interesting.  The macOS porttest failed despite the fix in #5632 being included.

File: compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp
Comment (fjeremic) [inline compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:66]: We should NULL initialize this.

File: compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp
Comment (fjeremic) [inline compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:178]: This assert should be improved to print the values of the two arguments. We should also use the `_WITH_NODE` assert here as well.

File: compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp
Comment (fjeremic) [inline compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:181]: `reinterpret_cast` need here as well as elsewhere. Will stop commenting on this one.

File: compiler/x/objectfmt/FunctionCallData.hpp
Comment (fjeremic) [inline compiler/x/objectfmt/FunctionCallData.hpp:74]: I guess this can be subjective, but to me the `TR::CodeGenerator*` is _always_ needed for these objects to be constructed properly in each overload of this constructor. Where as the other parameters may change. To me it makes most sense to put all the parameters which are consistently _always_ present first. Given these APIs are exclusively used in the codegen, we should just put the `TR::CodeGenerator*` parameter first.

**Review by fjeremic — CHANGES_REQUESTED** [review]

File: compiler/objectfmt/FunctionCallData.hpp
Comment (r30shah) [inline compiler/objectfmt/FunctionCallData.hpp:35]: Is the FunctionCallData needed to be short lived object only? Do we need `TR_ALLOC` for the FunctionCallData in case we want to cache the information? One example, I was thinking and currently trying out is in Snippet API where as I mentioned in https://github.com/eclipse/omr/pull/5575/files#r517395961, We can add `print` API and for that if we have cached the function call data to snippet, it will be easier.

File: compiler/codegen/OMRCodeGenerator.hpp
Comment (ymanton) [reply compiler/codegen/OMRCodeGenerator.hpp:1460]: Agree. Let's just stick with ObjectFormat, short names are hard to guess if you don't already know they exist.

File: compiler/objectfmt/FunctionCallData.hpp
Comment (dsouzai) [reply compiler/objectfmt/FunctionCallData.hpp:35]: > Do we need TR_ALLOC for the FunctionCallData in case we want to cache the information?

All `TR_ALLOC` does is add a bunch of `placement new` definitions. You don't technically need it since you _could_ just do something like
```
void *storage = allocateHeapMemory(sizeof(FunctionCallData));
FunctionCallData *fcData = new (storage) FunctionCallData(...);
```
Obviously this is terrible code, but the point being that `TR_ALLOC` doesn't give you any new caching ability, just that it allows cleaner code, ie
```
FunctionCallData *fcData = new (comp->trHeapMemory()) FunctionCallData(...);
```

Comment (r30shah) [reply compiler/objectfmt/FunctionCallData.hpp:35]: Okk yes, We do not need `TR_ALLOC` for allocating object on the heap. So let me rephrase my question, can we add it to list of object types in TR_Memory so that it can be used in more cleaner code as mentioned by @dsouzai  in previous comment?

File: compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp
Comment (Leonardo2718) [inline compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:189]: Perhaps I'm not familiar enough with this area of the code, but this looks a bit unintuitive to me. Every other occurrence of this pattern passes `data.pervInstr` to `generate*Instruction`, but here we pass `data.out_loadInstr`. Maybe a comment explaining the reason would help?

File: compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp
Comment (Leonardo2718) [inline compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:206]: This can be a const.

File: compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp
Comment (Leonardo2718) [inline compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:207]: Personally, I find this kind of combined dereference and increment a bit difficult to read.

File: compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp
Comment (Leonardo2718) [inline compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:209]: This too can be a const.

File: compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp
Comment (Leonardo2718) [inline compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:210]: This should b a reinterpret_cast.

File: compiler/objectfmt/JitCodeRXObjectFormat.hpp
Comment (Leonardo2718) [inline compiler/objectfmt/JitCodeRXObjectFormat.hpp:35]: I don't think this constructor declaration is needed.

**Comment by 0xdaryl** [comment]
I simply rebased and pushed to resolve the merge conflict.  I am working on addressing the comments now.

File: compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp
Comment (Leonardo2718) [inline compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:37]: Maybe this should be NULL initialized to be safe.

File: compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp
Comment (Leonardo2718) [inline compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:75]: Maybe this should be NULL initialized to be safe.

File: compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp
Comment (Leonardo2718) [inline compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:110]: Maybe this should be NULL initialized to be safe.

File: compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp
Comment (Leonardo2718) [inline compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:136]: Maybe this should be NULL initialized to be safe.

File: compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp
Comment (Leonardo2718) [inline compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:172]: This can be a const.

File: compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp
Comment (Leonardo2718) [inline compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:180]: I think this can be a const.

**Comment by 0xdaryl** [comment]
@r30shah @fjeremic @Leonardo2718 : Thanks for your careful reviews and sorry for the delay in responding.  I believe I have addressed all the comments that were raised.  I would appreciate a second review to ensure your comments or questions were addressed.  Thank you!

File: compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp
Comment (Leonardo2718) [reply compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:214]: Fair enough.

File: compiler/objectfmt/OMRFunctionCallData.hpp
Comment (fjeremic) [reply compiler/objectfmt/OMRFunctionCallData.hpp:120]: We should likely make this a `struct` then? Minor detail I suppose. `struct` fields are public by default where as `class` fields are private by default.

**Review by fjeremic — APPROVED** [review]
Looks reasonable to me with all the new updates addressing previous concerns.

**Comment by 0xdaryl** [comment]
@genie-omr build all

File: compiler/objectfmt/OMRFunctionCallData.hpp
Comment (0xdaryl) [reply compiler/objectfmt/OMRFunctionCallData.hpp:120]: Yes, you're right.  But I left it as a class because it is decorated `OMR_EXTENSIBLE`.  I didn't think through all the implications of making "extensible structs".  In theory we shouldn't have a problem (maybe some work on the linter?) but perhaps something to be solved in a different PR.

**Comment by 0xdaryl** [comment]
If everyone is ok with this, could a committer please merge?

For additional integration testing, I have tested it builds successfully on all platforms with the latest OpenJ9 master.

File: compiler/x/objectfmt/FunctionCallData.hpp
Comment (Leonardo2718) [reply compiler/x/objectfmt/FunctionCallData.hpp:45]: Ok, that's fine.

My main concern here is that making sure the correct overload is picked can be error-prone, so having a descriptive name would disambiguate which version the programmer intended to call. Also, it's unclear right now what invariants each constructor is enforcing.

**Comment by Leonardo2718** [comment]
LGTM :+1: 

@fjeremic you're the assigned committer on this one :wink:

**Pull request merged**
